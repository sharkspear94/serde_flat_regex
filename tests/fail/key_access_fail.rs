use serde::Deserialize;
use serde_flat_regex::flat_regex;
use std::ffi::CString;

#[flat_regex]
#[derive(Debug, Deserialize, PartialEq, Clone)]
struct Foo<'a> {
    id: u32,
    #[flat_regex(
        regex = r"lanport(status|speed)_\d+",
        key_access = "some_modul::as_str"
    )]
    #[serde(borrow)]
    rest: std::collections::HashMap<CString, &'a str>,
}

pub mod some_modul {
    use std::str::Utf8Error;

    pub fn as_str(s: &String) -> Result<&str, Utf8Error> {
        Ok(s.as_str())
    }
}

fn main() {}
