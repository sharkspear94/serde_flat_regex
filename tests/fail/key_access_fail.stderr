error[E0308]: mismatched types
  --> tests/fail/key_access_fail.rs:5:1
   |
5  | #[flat_regex]
   | ^^^^^^^^^^^^^ expected struct `std::string::String`, found struct `CString`
...
11 |         key_access = "some_modul::as_str"
   |                      -------------------- arguments to this function are incorrect
   |
   = note: expected reference `&std::string::String`
              found reference `&CString`
note: function defined here
  --> tests/fail/key_access_fail.rs:20:12
   |
20 |     pub fn as_str(s: &String) -> Result<&str, Utf8Error> {
   |            ^^^^^^ ----------
   = note: this error originates in the attribute macro `flat_regex` (in Nightly builds, run with -Z macro-backtrace for more info)
